yt_dlp/extractor/camsoda.py

```diff
@@ -0,0 +1,63 @@
+import random
+
+from .common import InfoExtractor
+from ..utils import ExtractorError, traverse_obj
+
+
+class CamsodaIE(InfoExtractor):
+    _VALID_URL = r'https?://www\.camsoda\.com/(?P<id>[0-9A-Za-z-]+)'
+    _TESTS = [{
+        'url': 'https://www.camsoda.com/lizzhopf',
+        'info_dict': {
+            'id': 'lizzhopf',
+            'ext': 'mp4',
+            'title': 'lizzhopf (lizzhopf) Nude on Cam. Free Live Sex Chat Room - CamSoda',
+            'description': str,
+            'is_live': True,
+            'age_limit': 18,
+        },
+        'skip': 'Room is offline',
+    }]
+
+    _ROOM_OFFLINE = 'Model is offline.'
+    _ROOM_PRIVATE = 'Model is in private show.'
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        user_id = 'guest_%u' % random.randrange(10000, 99999)
+        webpage = self._download_webpage(url, video_id, headers=self.geo_verification_headers())
+
+        data = self._download_json(
+            f'https://camsoda.com/api/v1/video/vtoken/{video_id}?username={user_id}',
+            video_id, headers=self.geo_verification_headers())
+        if not data:
+            raise ExtractorError('Unable to find configuration for stream.')
+
+        private_servers = traverse_obj(data, ('private_servers'), expected_type=list)
+        if private_servers:
+            raise ExtractorError(self._ROOM_PRIVATE, expected=True)
+
+        stream_name = traverse_obj(data, ('stream_name'), expected_type=str)
+        if not stream_name:
+            raise ExtractorError(self._ROOM_OFFLINE, expected=True)
```

Comment:
shouldnt these checks be done only if there is no format?

Reply:
If the room is offline, there server data isn't available in the response, so I thought checking these conditions first would make more sense.

Online response:

```json
{
    "aspect": 16,
    "edge_servers": [
        "edge54-ord.livemediahost.com",
        "edge55-ord.livemediahost.com"
    ],
    "ffmpeg_server": "ded4589",
    "height": 1080,
    "status": 1,
    "stream_name": "cam_obs/catydance-flu",
    "token": "[REDACTED]",
    "width": 1920
}
```

Offline response:

```json
{
    "aspect": 16,
    "edge_servers": [],
    "ffmpeg_server": null,
    "height": 1080,
    "status": 1,
    "stream_name": "",
    "token": "[REDACTED]",
    "width": 1920
}
```

Private show response:

```json
{
    "aspect": 16,
    "edge_servers": [],
    "ffmpeg_server": "ded3441",
    "height": 540,
    "status": 1,
    "stream_name": "1280x720/agataruiz-flu-ingest7-bog",
    "token": "[REDACTED]",
    "width": 960,
    "private_servers": [
        "ingest7-bog.livemediahost.com"
    ]
}
```